src/Authorization/FitOnFhir.Authorization/Services/TokenValidationService.cs

```diff
@@ -0,0 +1,158 @@
+                _logger.LogError("The request Authorization header is invalid.");
+                return false;
+            }
+
+            // Read the token into the handler.
+            _jwtSecurityTokenHandlerProvider.SetMapInboundClaims(false);
+            JwtSecurityToken jwtSecurityToken;
+            try
+            {
+                jwtSecurityToken = _jwtSecurityTokenHandlerProvider.ReadJwtToken(token);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "The request JWT is malformed.");
+                return false;
+            }
+
+            // Wait, if necessary, for all metadata configuration to be returned
+            Task t = Task.WhenAll(_configurationTasks);
+            await t.WaitAsync(cancellationToken);
+
+            // Find the correct authority from the list
+            if (!_issuers.TryGetValue(jwtSecurityToken.Issuer, out var authority))
+            {
+                _logger.LogInformation("Issuer {0} not found in the list of authorized identity providers.", jwtSecurityToken.Issuer);
+                return false;
+            }
+
+            // Fetch the signing keys from the authority.
+            var config = await _openIdConfigurationProvider.GetConfigurationAsync(authority, cancellationToken);
+
+            try
+            {
+                var tokenValidationParameters = new TokenValidationParameters
+                {
+                    RequireSignedTokens = true,
+                    ValidAudience = _authenticationConfiguration.Audience,
+                    ValidateAudience = true,
+                    ValidIssuer = config.Issuer,
+                    ValidateIssuer = true,
+                    ValidateIssuerSigningKey = true,
+                    ValidateLifetime = true,
+                    IssuerSigningKeys = config.SigningKeys,
+                };
+
+                // Validate the token.
+                var tokenValidationResult = await _jwtSecurityTokenHandlerProvider.ValidateTokenAsync(jwtSecurityToken.RawData, tokenValidationParameters);
+
+                return tokenValidationResult.IsValid;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Failed to validate the request Bearer token.");
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Creates a mapping between the metadata endpoints provided in <see cref="AuthenticationConfiguration"/>.AuthorizedIdentityProviders
+        /// and the name of the issuer, as declared in the <see cref="OpenIdConnectConfiguration"/> Issuer property for that endpoint's config.
+        /// This mapping can be used to determine which endpoint to authenticate tokens against, when a user wishes to authorize access to a fitness data provider.
+        /// Configuration data requests are made asynchronously.
+        /// </summary>
+        private void CreateIssuerMapping()
+        {
+            try
```

Comment:
I think this try/catch needs to be done inside the loop so that the exception for each providerEndpoint is captured. We could also write the providerEndpoint to the error message so that there is more data about the failure.